packages/components/switch/src/switch.ts

```diff
@@ -0,0 +1,164 @@
+  validationStyles
+} from '@sl-design-system/shared';
+import type { IconSize } from '@sl-design-system/icon';
+import { LitElement, html } from 'lit';
+import { property, queryAssignedElements } from 'lit/decorators.js';
+import styles from './switch.scss.js';
+
+export type SwitchSize = 'sm' | 'md' | 'lg';
+export type SwitchOrientation = 'horizontal' | 'vertical';
+
+export class Switch extends FormControlMixin(HintMixin(LitElement)) {
+  /** @private */
+  static formAssociated = true;
+
+  /** @private */
+  static override styles: CSSResultGroup = [validationStyles, styles];
+
+  #events = new EventsController(this);
+
+  #validation = new ValidationController(this, {
+    validators: [requiredValidator]
+  });
+  #initialState = false;
+
+  /** Element internals. */
+  readonly internals = this.attachInternals() as ElementInternals & IElementInternals;
+
+  /** Emits when the checked state changes. */
+  @event() change!: EventEmitter<boolean>;
+
+  /** Whether the switch is on or off. */
+  @property({ type: Boolean, reflect: true }) checked?: boolean;
+
+  /** Whether the checkbox is invalid. */
+  @property({ type: Boolean, reflect: true }) invalid?: boolean;
+
+  /** Custom icon in "off" state. */
+  @property({ reflect: true }) iconOff?: string;
+
+  /** Custom icon in "on" state. */
+  @property({ reflect: true }) iconOn?: string;
+
+  /** Switch size. */
+  @property({ reflect: true }) size: SwitchSize = 'md';
+
+  /** Switch orientation. */
+  @property({ reflect: true }) orientation: SwitchOrientation = 'horizontal';
+
+  /** The value for the switch. */
+  @property() value?: string;
+
+  @queryAssignedElements() label?: string;
+
+  get icon(): string {
+    return this.checked ? this.iconOn || 'check' : this.iconOff || 'xmark';
+  }
+
+  get iconSize(): IconSize {
+    return this.size === 'md' ? 'xs' : 'md';
+  }
+
+  override connectedCallback(): void {
+    super.connectedCallback();
+
+    this.internals.role = 'switch';
+    this.internals.ariaChecked = this.checked ? 'true' : 'false';
+
+    this.#events.listen(this, 'click', this.#onClick);
+    this.#events.listen(this, 'keydown', this.#onKeydown);
+
+    this.setFormControlElement(this);
+
+    this.#validation.validate(this.checked ? this.value : undefined);
+
+    if (!this.hasAttribute('tabindex')) {
+      this.tabIndex = 0;
+    }
+  }
+
+  override updated(changes: PropertyValues<this>): void {
+    super.updated(changes);
+
+    if (changes.has('checked')) {
+      this.internals.ariaChecked = this.checked ? 'true' : 'false';
+
+      if (this.checked && this.internals.states) {
+        this.internals.states.add('--checked');
+      } else {
+        this.internals.states.delete('--checked');
+      }
+    }
+
+    if (changes.has('checked') || changes.has('value')) {
+      this.setFormValue(this.checked ? this.value : undefined);
+    }
+  }
+
+  formAssociatedCallback(): void {
+    this.#initialState = this.getAttribute('checked') === null ? false : true;
```

Comment:
Maybe we can simplify it a little bit to `this.#initialState = this.getAttribute('checked') !== null` ?